megatron/neox_arguments/arguments.py

```diff
@@ -0,0 +1,748 @@
+        a number of function are performed in order to 
+        calculate values and  assert consistency
+        """
+        if not NeoXArgs.validate_keys():
+            raise ValueError(self.__class__.__name__+".__post_init__() NeoXArgs keys cannot be validated")
+
+        self.enable_logging()
+
+        if not self.only_default_arguments:
+            self.configure_distributed_args()
+            self.calculated_derived()
+        
+        if not self.validate_types():
+            raise ValueError(self.__class__.__name__+".__post_init__() NeoXArgs types cannot be validated")
+
+        if not self.only_default_arguments:
+            if not self.validate_values():
+                raise ValueError(self.__class__.__name__+".__post_init__() NeoXArgs values cannot be validated")
+            
+        
+        self.save_yml()
+
+    @classmethod
+    def from_ymls(cls, paths_to_yml_files: List[str], overwrite_values=None):
+        """
+        instantiates NeoXArgs while reading values from yml files
+        """
+
+        print(cls.__name__+".from_ymls() "+str(paths_to_yml_files), flush=True)
+
+        # initialize an empty config dictionary to be filled by yamls
+        config = dict()
+
+        # iterate of all to be loaded yaml files
+        for conf_file_name in paths_to_yml_files:
+
+            # load file
+            with open(conf_file_name) as conf_file:
+                conf = yaml.load(conf_file, Loader=yaml.FullLoader)
+
+            # check for key duplicates and load values
+            for conf_key, conf_value in conf.items():
+                if conf_key in config:
+                    raise ValueError(f'Conf file {conf_file_name} has the following duplicate keys with previously loaded file: {conf_key}')
+
+                conf_key_converted = conf_key.replace("-", "_")  #TODO remove replace and update configuration files?
+                config[conf_key_converted] = conf_value
+
+        #TODO check for unspecified params?
+        # old code:
+        # # Configuration parameters not specified
+        # params_missing = [key for key in ds_runner_keys + megatron_keys + ds_config_keys + neox_config_keys
+        #                   if key not in conf]
+        # if len(params_missing) > 0:
+        #     log.debug(f'Configuration parameters not specified: {", ".join(params_missing)}')
+
+        if overwrite_values is not None:
+            for k, v in overwrite_values.items():
+                config[k] = v
+
+        # instantiate class and return
+        # duplicate values and unrecognized keys are again checked upon instantiation
+        return cls(**config)
+
+    @classmethod
+    def from_dict(cls, args_dict: dict):
+        """
+        instantiates NeoXArgs while reading values from input dict
+        """
+
+        print(cls.__name__+".from_dict() ", flush=True)
+
+        return cls(**args_dict)
+
+    def update_value(self, key: str, value):
```

Comment:
You can set any variable name, also when not defined in NeoXArgs. I see the following advantages to check that an attribute is defined in NeoXArgs:

- You can avoid typos. Setting a variable name to e.g. "args.bacth_size = 42" would not run into an error. You are still likely to use "args.batch_size" in the code. In fact autocomplete would suggest to. The update_value function would raise an error and save you a potentially lengthy debugging. I used in the the refacturing process to avoid errors like this.
- Checking for types only works when the attribute and type are defined.
- NeoXArgs would actually be a representative overview of all configs in the code.

=> no harm, optional to use, advantages when used